### PR TITLE
[generate] Opt-in normalized repetition penalty

### PR DESCRIPTION
**Pull request by captainpete**
[CPU CI](https://transformers-ci.lor-e.huggingface.cool/d/pytest-observability-by-pr/pytest-observability-branch?var-pr=48642) [GPU run-slow](https://transformers-ci.lor-e.huggingface.cool/d/pytest-observability-by-pr/pytest-observability-branch?var-pr=48642)

# What does this PR do?

The `repetition_penalty` currently applies two different operators depending on decoding strategy. Beam search has applied its logits-processor chain to log-probabilities since at least v4.0.0; greedy and sampled decoding apply it
to raw logits. This adds a normalized opt-in variant to greedy/sampled consistent with the beam path.

Follow-on from #47595 (discussed there; @qgallouedec [invited the PR](https://github.com/huggingface/transformers/issues/47595#issuecomment-5397554853)).

## Analysis

The [arXiv note](https://arxiv.org/abs/2607.09791v2) and [analysis code](https://github.com/captainpete/repetition-penalty-gauge) contain the full details. I've run a few additional experiments (the matched calibration battery) and pushed v2 of the note since I posted the original issue.

The TL;DR from the paper: The raw form branches on the sign of each logit, `torch.where(score < 0, score * penalty, score / penalty)`, and because softmax is invariant to adding a constant to every logit, that sign is an arbitrary coordinate across different models: shifting all logits by a constant is a no-op for the model, yet under the penalty it rewrites 87–93% of greedy tokens at the deployed θ=1.1 anchor (three 2.8B–7B models). At θ=1.3 the raw operator drops valid, schema-conformant output on 200 JSONSchemaBench schemas from 97% to 23%.

The sign-branch change that introduced the weirdness made its way to essentially every downstream inference engine. I figured I'd start here first and roll out subsequent PRs depending on how this goes.

## What this PR adds

This PR adds an opt-in `normalize` mode to `RepetitionPenaltyLogitsProcessor` and `EncoderRepetitionPenaltyLogitsProcessor`, exposed as `repetition_penalty_normalize` and `encoder_repetition_penalty_normalize` in `GenerationConfig`. With the flag on, the penalty is applied to log-probabilities; every value is then ≤ 0 and the operator is well-defined.

Keeping with project conventions, downstream processors continue to receive logits under greedy decoding and sampling.

## Tests

```
python -m pytest tests/generation/test_logits_process.py tests/generation/test_configuration_utils.py
# 89 passed, 5 skipped
ruff check src/transformers/generation tests/generation && ruff format --check src/transformers/generation tests/generation
python utils/check_docstrings.py
# both clean
```

## Disclosure on AI use

As this is my first contribution to the repo, I dusted off the keyboard, cracked open nvim and tig, cut a new branch, and fully implemented this myself. Please don't judge the PR based on my editor choices, I was raised in a Vim household.

Now for the research paper itself I was using Claude Code, mostly Fable, with a whole deck of skills that I use on different projects. So I had an AI-authored branch with an implementation before I even thought about writing a contribution here. I used that branch as a reference to make sure this PR reflects the paper results. Both approaches ended up near identical on the inference path, and I reproduced the paper's A1 gauge result on gpt2 through this branch's flag: same 200 prefixes and model revision, token-identical streams to the paper's harness, flip rate 0.58 raw and zero normalized.

What didn't end up coming across were more tests, and some additional config-sanity guard-clauses. I figure we can add those if we want the bells and whistles. Keen to understand if the core change makes sense first.

## Coordination

The change was proposed on #47595 with the paper, a runnable repro, and this branch.

## Differentiation

No open PR references #47595. #48468 touches the same class (it flips `supports_continuous_batching`) but is orthogonal; the continuous-batching branch here already handles `normalize`.

## Before submitting

- [x] (First-time contributors only) I confirm that this PR description and code is not written by an LLM or code agent. Yes, with Disclosure.
- [x] I read the contributor guideline and the PR checks.
- [x] Discussed via GitHub issue: yes, #47595 linked above.
- [x] Documentation: docstrings for the new arguments and config fields (picked up by the existing autodoc entries; no separate docs page lists these flags).
- [x] Tests added.

## Who can review?

@Cyrilvallez @zucchini-nlp (routing suggested by @qgallouedec on the issue).

### CODE REVIEW

File: src/transformers/generation/logits_process.py [hunk 1]
```diff
@@ -303,6 +303,15 @@ def __call__(self, input_ids: torch.LongTensor, scores: torch.FloatTensor) -> to
         return scores_processed
 
 
+def _apply_penalty(score: torch.FloatTensor, penalty: float, normalize: bool) -> torch.FloatTensor:
+    if normalize:
+        # score is log-probs, always negative
+        return score * penalty
+    else:
+        # score is raw logits, positive values have to be divided to move them the right direction
+        return torch.where(score < 0, score * penalty, score / penalty)
```
Comment (Cyrilvallez) [inline src/transformers/generation/logits_process.py:312]: I'm very confused, this is exactly the same either way

File: src/transformers/generation/logits_process.py [hunk 2]
```diff
@@ -387,28 +402,32 @@ def __call__(self, input_ids: torch.LongTensor, scores: torch.FloatTensor) -> to
...
-                penalty_scores = torch.where(last_scores < 0, last_scores * self.penalty, last_scores / self.penalty)
+                penalty_scores = _apply_penalty(last_scores, self.penalty, self.normalize)
                 scores[:, -1, :] = torch.where(token_mask, penalty_scores, last_scores)
             return scores
 
         if input_ids.dim() == 1:
             input_ids = input_ids.unsqueeze(1)
 
+        if self.normalize:
+            scores = torch.log_softmax(scores, dim=-1)
```
Comment (Cyrilvallez) [inline src/transformers/generation/logits_process.py:427]: Any downstream processors will see log softmax of the scores

**Review by Cyrilvallez — COMMENTED** [review]
I'm quite confused by the PR, as it makes changes based on `normalize` to ops that already do exactly the same. Also, returning log softmax means every other downstream processor will see log softmax of the scores instead of the logits

File: src/transformers/generation/logits_process.py (hunk 1 above)
Comment (qgallouedec) [reply src/transformers/generation/logits_process.py:312]: Yep, this helper adds nothing: on log-probs the `where` is already a multiply, so the only change is the `log_softmax`, and the PR should say so and drop the helper

Comment (captainpete) [reply src/transformers/generation/logits_process.py]: Ah. Yes that makes sense. That's what I get for cementing an unneeded conditional in my head then trying to clean it up later. I have removed the helper (758c94257599fbe5e5870db8349e501571fe49ac). It's a much better diff now, just the `log_softmax`.

File: src/transformers/generation/logits_process.py (hunk 2 above)
Comment (captainpete) [reply src/transformers/generation/logits_process.py]: Yes. My precedent for this was checking that beam emits log-probs to downstream processors. The built-in processors get used in both cases. I'll look more into this today if I get a chance, I'd like to confirm a couple of things.

Comment (captainpete) [reply src/transformers/generation/logits_process.py]: Thanks for the review and patience. Reviewing again I can see the contract is spelled out pretty clearly in the scores docstring: log-softmax for beam, logits otherwise.

Looking at the built-in processors. They all look shift-invariant except for `ExponentialDecayLengthPenalty` which uses `torch.abs(scores[:, eos_token_id])` without any special handling for log-probs, so changing greedy/beam when using that processor has an effect independent of this PR.

I'll have a go at restoring the logit scale. The math works out, but I haven't finished coding or testing it yet.

Comment (captainpete) [reply src/transformers/generation/logits_process.py]: Okay, logits restored in 798830c.

This version uses the `logsumexp` trick to shift the rows into log-prob space, applies the existing `where`, then shifts back. Same penalty and sampling distribution as before but downstream now gets logits.

A bit more green in the diff since it has to keep track of the scale, but it turned out pretty straightforward.

Verified against the paper harness with the same results as the previous version.
Tightened up some of the tests now we can have exact matches on unmodified tokens.

I'll leave it there for the minute. Let me know what you think. Thanks again for the review.

File: src/transformers/generation/logits_process.py [hunk 3]
```diff
@@ -415,8 +433,13 @@ def __call__(self, input_ids: torch.LongTensor, scores: torch.FloatTensor) -> to
         scores = torch.nn.functional.pad(scores, (0, 1))
 
         score = torch.gather(scores, 1, input_ids)
+        if self.normalize:
+            lse = torch.logsumexp(scores[..., :vocab_size], dim=-1, keepdim=True)
+            score = score - lse
```
Comment (Cyrilvallez) [inline src/transformers/generation/logits_process.py:438]: Here you don't know if the scores are logits or log probs already...

**Review by Cyrilvallez — COMMENTED** [review]
Hey, sorry but for all the reasons I mentioned before and now, I'm not convinced by the PR!

File: src/transformers/generation/logits_process.py (hunk 3 above)
Comment (captainpete) [reply src/transformers/generation/logits_process.py:438]: Yes, there's no way to tell. It's functionally the same both ways though: `lse == 0` for log-probs, for logits it is the row's log-normalizer, so `score - lse` is always the log-prob at this point.

**Comment by captainpete** [comment]
Thanks for taking another look. I'll grant the point isn't obvious from the diff.
I'm convinced this is an issue, I'm not convinced this is the best solution, keen to get your direction.

#### Is this an issue?

Yes, in 2019 the `where` fix arrived in #2302 / #2303 as "let's go for this solution for now", [with caveats](https://github.com/huggingface/transformers/pull/2303#issuecomment-587962571). It was then copied by [vLLM](https://github.com/vllm-project/vllm/pull/1424), TGI, LMDeploy, [mlx-lm](https://github.com/ml-explore/mlx-examples/pull/399), and [ExLlamaV2](https://github.com/turboderp-org/exllamav2/blob/7dc12af3a81f34ac3f27cd7602ed539b638933ca/exllamav2/exllamav2_ext/cpp/sampling.cpp#L85-L86). Llama.cpp derived the same fix [with concerns](https://github.com/ggml-org/llama.cpp/issues/2970).

Current models trigger the original caveats. The issue shows up as rep penalty [misbehaving on structured output](https://github.com/vllm-project/vllm/issues/40080). The paper details how it all works. Raised here because maintainers turn to HF for a verdict [whenever](https://github.com/vllm-project/vllm/pull/1424) [they](https://github.com/vllm-project/vllm/issues/1620) [need](https://github.com/vllm-project/vllm/discussions/539) [precedent](https://github.com/ml-explore/mlx-examples/pull/399).

#### Is this the best solution?

Mathematically it's good, implementation-wise probably not.

The minimal fix is processing on log-probs, which is what beam already does; this PR takes a stab at an opt-in for greedy/sampling, and now preserves logits per your direction. But! It also adds configs and modifies a processor that crystalized back in 2019. If HF has an "upstream solution" to this, what would it look like? A few options I think:

0. Users find this PR/issue, use the current workarounds (beam/tune-where-possible/avoid).
1. Keep this PR sans config, so `logits_processor=[RepetitionPenaltyLogitsProcessor(1.3, normalize=True)]`
2. A separate `NormalizedRepetitionPenaltyLogitsProcessor`, probably cleanest, but it's a new thing
3. Off-core: a `custom_generate` repo like the [ones moved out of core](https://huggingface.co/collections/transformers-community/custom-generation-methods-tutorials-6823589657a94940ea02cfec), with a link in the docstring for users that want the normalized variant on the greedy/sampling path. Or I'll host it on my GH and link it here before we close out if you prefer.

I'll jump in whichever direction is the preference, please let me know and thanks again for taking a look at this.
